car-bridge/src/main.rs

```diff
@@ -29,58 +41,200 @@ async fn main() -> Result<(), Box<dyn Error>> {
 
     let vin = env::<String>(VIN_ENV_NAME);
     let vin = vin.as_deref().unwrap_or(DEFAULT_VIN);
-    let host = env::<String>(BROKER_URL_ENV_NAME);
-    let host = host.as_deref().unwrap_or(DEFAULT_BROKER_URL);
-
-    info!("Connecting to MQTT broker on '{host}'.");
-
-    let mut client = AsyncClient::new(
-        CreateOptionsBuilder::new()
-            .mqtt_version(MQTT_VERSION_5)
-            .server_uri(host)
-            .client_id(format!("car-bridge-{vin}"))
-            .finalize(),
-    )?;
-
-    let mut message_stream = client.get_stream(MQTT_CLIENT_BUFFER_SIZE);
 
-    client.connect(ConnectOptionsBuilder::new().mqtt_version(MQTT_VERSION_5).finalize()).await?;
+    let chariott = GrpcChariott::connect().await?;
 
-    let c2d_topic = format!("c2d/{vin}");
-    info!("Subscribing to topic '{c2d_topic}'.");
-    client.subscribe(c2d_topic.clone(), QOS_2).await?;
-
-    // Publish a phony message.
-    client.publish(Message::new(c2d_topic, "Example payload", QOS_2)).await?;
+    let mut client = MqttMessaging::connect(vin.to_owned()).await?;
+    let mut messages = client.subscribe(format!("c2d/{vin}")).await?;
 
     let cancellation_token = ctrl_c_cancellation();
 
+    let (response_sender, mut response_receiver) = mpsc::channel(PUBLISH_BUFFER);
+
+    let publish_handle = {
+        // Detach sending the responses from handling the messages to avoid
+        // backpressure and disconnect the client gracefully.
+
+        let cancellation_token = cancellation_token.child_token();
+        spawn(async move {
+            let client = Arc::new(client);
+
+            loop {
+                select! {
+                    message = response_receiver.recv() => {
+                        let Some((topic, message)) = message else {
+                            warn!("Response receiver stopped, no more messages will be published.");
+                            break;
+                        };
+
+                        {
+                            let client = Arc::clone(&client);
+
+                            spawn(async move {
+                                if let Err(e) = client.publish(topic, message).await {
+                                    debug!("Error when publishing message: '{:?}'.", e);
+                                }
+                            });
+                        }
```

Comment:
remove redundant scope:

```suggestion
                        let client = Arc::clone(&client);

                        spawn(async move {
                            if let Err(e) = client.publish(topic, message).await {
                                debug!("Error when publishing message: '{:?}'.", e);
                            }
                        });
```